include/bluetooth/audio/hap.h

```diff
@@ -0,0 +1,275 @@
+ *
+ * @param _name Name of callback structure.
+ */
+#define BT_HAP_CB_DEFINE(_name)							\
+	static const STRUCT_SECTION_ITERABLE(bt_hap_cb,				\
+					     _CONCAT(bt_hap_cb_, _name))
+
+/**
+ * @brief Create profile instance to remote Hearing Access Service
+ *
+ * This will start a GATT discovery to setup handles and subscriptions.
+ * This shall be called once before any other actions can be executed for the
+ * peer device, and the @ref bt_hap_cb connected callback will notify
+ * when it is possible to start remote operations.
+ * The profile instance is removed when the bond is removed or when @ref
+ * bt_hap_delete is called.
+ * 
+ * @param conn The connection to initialize the profile for.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_new(struct bt_conn *conn);
+
+/**
+ * @brief Delete profile instance
+ *
+ * This will remove stored profile instance of remote Hearing Aid Service.
+ *
+ * @param hap Profile instance object.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_delete(struct bt_hap *hap);
+
+/**
+ * @brief Iterate through all existing profile instances.
+ *
+ * Iterate profile instances of remote Hearing Access Services.
+ *
+ * @param func Function to call for each profile instance.
+ * @param user_data Data to pass to the callback function.
+ */
+void bt_hap_foreach(bool (*func)(const struct bt_hap *hap, void *user_data),
+		    void *user_data);
+
+/**
+ * @brief Get profile instance to local Hearing Access Service.
+ *
+ * Get the profile instance connection to local Hearing Access Service.
+ *
+ * @return Profile instance object or NULL if there is no service.
+ */
+struct bt_hap *bt_hap_local_get(void);
+
+/**
+ * @brief Get the Bluetooth connection object of the profile instance.
+ *
+ * Get the Bluetooth connection object of a HAP instance.
+ *
+ * @param hap Profile instance object.
+ *
+ * @return Bluetooth connection object or NULL.
+ */
+struct bt_conn *bt_hap_conn_get(const struct bt_hap *hap);
+
+/** 
+ * @typedef bt_hap_complete_func_t
+ * @brief Operation complete callback.
+ *
+ * @param hap Profile instance object.
+ * @param err Error code.
+ * @param user_data User data passed to callback.
+ */
+typedef void (*bt_hap_complete_func_t) (struct bt_hap *hap, int err,
+					void *user_data);
+
+/**
+ * @brief Write Preset Name.
+ *
+ * Write the Name of a writable preset record identified by the Index.
+ *
+ * @param hap Profile instance object.
+ * @param index Index of preset record that name to be written.
+ * @param data Name to be written.
+ * @param length Name length.
+ * @param func Operation complete callback.
+ * @param user_data User data to be passed back to callback.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_preset_write_name(struct bt_hap *hap, uint8_t index,
+			     const void *data, uint8_t length,
```

Comment:
Given that this is a specific function to write a specific type of value (name), I suggest to use
```suggestion
			     const char *name,
```

instead. We can then internally use `strlen` to get the length.

Reply:
The string shall be null-terminated then, thus it's more safe to request length from the user.

Reply:
> The string shall be null-terminated then

Yes, but that's fair just to put in the documentation. See e.g. `bt_set_name`. 

> thus it's more safe to request length from the user.

If you cannot trust the user to NULL-terminate a string, you cannot trust the user to put the right length either IMO. Doing value & length instead of a NULL terminated string are equally dangerous, but the string (`char *`) API is more semantic as you are in fact writing a string.

Reply:
OK, I'll change that ^^